

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Genotype Imputation &mdash; GWASpy 0.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/custom.css?v=eea1f72d" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=01f34227"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorial" href="tutorial.html" />
    <link rel="prev" title="Haplotype Phasing" href="phasing.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            GWASpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="qb.html">Hail Query and Batch</a></li>
<li class="toctree-l1"><a class="reference internal" href="preimp_qc.html">Pre-Imputation QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="pca.html">Principal Component Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="phasing.html">Haplotype Phasing</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Genotype Imputation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#examples">Examples</a></li>
<li class="toctree-l2"><a class="reference internal" href="#arguments-and-options">Arguments and options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#output">Output</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorial.html">Tutorial</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GWASpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Genotype Imputation</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/imputation.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="genotype-imputation">
<span id="sec-imputation"></span><h1>Genotype Imputation<a class="headerlink" href="#genotype-imputation" title="Link to this heading"></a></h1>
<p>Genotype imputation is a process of estimating missing genotypes from the haplotype or genotype reference panel. It
allows you to accurately evaluate the evidence for association at genetic markers that are not directly genotyped.
GWASpy has a module, <code class="code docutils literal notranslate"><span class="pre">imputation</span></code>, for running imputation using IMPUTE5. Because imputation can be a computationally
intensive task, we run it on multiple chunks in parallel, then merge the imputed chunks together at the end. Below are
examples of how to run imputation using either the HGDP+1kGP or your own reference panel</p>
<section id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Link to this heading"></a></h2>
<p><strong>1. HGDP+1kGP reference panel</strong></p>
<blockquote>
<div><div class="highlight-sh notranslate"><div class="highlight"><pre><span></span>imputation<span class="w"> </span>--input-file<span class="w"> </span>gs://path/to/file.vcf.bgz<span class="w"> </span>--vcf-ref<span class="w"> </span>hgdp1kgp<span class="w"> </span>--output-filename<span class="w"> </span>my_outfilename<span class="w"> </span>--out-dir<span class="w"> </span>gs://path/to/output/dir<span class="w"> </span>--n-samples<span class="w"> </span><span class="m">1989</span><span class="w"> </span>--n-ref-samples<span class="w"> </span><span class="m">4091</span><span class="w"> </span>--billing-project<span class="w"> </span>my-billing-project
</pre></div>
</div>
</div></blockquote>
<p><strong>2. Own reference panel</strong></p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">imputation</span> <span class="o">--</span><span class="nb">input</span><span class="o">-</span><span class="n">file</span> <span class="n">gs</span><span class="p">:</span><span class="o">//</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">file</span><span class="o">.</span><span class="n">vcf</span><span class="o">.</span><span class="n">bgz</span> <span class="o">--</span><span class="n">vcf</span><span class="o">-</span><span class="n">ref</span> <span class="n">gs</span><span class="p">:</span><span class="o">//</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">ref_panel</span><span class="o">/</span><span class="n">ALL</span><span class="o">.</span><span class="n">chrCNUMBER</span><span class="o">.</span><span class="n">vcf</span> <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">filename</span> <span class="n">my_outfilename</span> <span class="o">--</span><span class="n">out</span><span class="o">-</span><span class="nb">dir</span> <span class="n">gs</span><span class="p">:</span><span class="o">//</span><span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="nb">dir</span> <span class="o">--</span><span class="n">n</span><span class="o">-</span><span class="n">samples</span> <span class="mi">1989</span> <span class="o">--</span><span class="n">n</span><span class="o">-</span><span class="n">ref</span><span class="o">-</span><span class="n">samples</span> <span class="mi">4091</span> <span class="o">--</span><span class="n">billing</span><span class="o">-</span><span class="n">project</span> <span class="n">my</span><span class="o">-</span><span class="n">billing</span><span class="o">-</span><span class="n">project</span>
</pre></div>
</div>
</div></blockquote>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>When using your own reference panel, make sure that you use the CNUMBER placeholder in the filename passed to –vcf-ref</p>
</div>
</section>
<section id="arguments-and-options">
<h2>Arguments and options<a class="headerlink" href="#arguments-and-options" title="Link to this heading"></a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 23.1%" />
<col style="width: 76.9%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--input-file</span></code></p></td>
<td><p>Path to where the VCF for target genotypes paths is</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--vcf-ref</span></code></p></td>
<td><p>Reference panel file to use for imputation</p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--local</span></code></p></td>
<td><p>Type of service. Default is Service backend where jobs are executed on a multi-tenant compute cluster in Google Cloud</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--billing-project</span></code></p></td>
<td><p>Billing project to be used for the jobs</p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--n-samples</span></code></p></td>
<td><p>Number of target samples to be imputed. We use this to estimate resources for some of the jobs</p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--n-ref-samples</span></code></p></td>
<td><p>Number of reference samples. We use this to estimate resources for some of the jobs</p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--software</span></code></p></td>
<td><p>Software to use for phasing. Options: [<code class="code docutils literal notranslate"><span class="pre">beagle5</span></code>, <code class="code docutils literal notranslate"><span class="pre">impute5</span></code>]. Default is <code class="code docutils literal notranslate"><span class="pre">impute5</span></code></p></td>
</tr>
<tr class="row-odd"><td><p><code class="code docutils literal notranslate"><span class="pre">--output-filename</span></code></p></td>
<td><p>Output filename without file extension</p></td>
</tr>
<tr class="row-even"><td><p><code class="code docutils literal notranslate"><span class="pre">--out-dir</span></code></p></td>
<td><p>Path to where output files will be saved</p></td>
</tr>
</tbody>
</table>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading"></a></h2>
<p>The resulting output is a VCF file per chromosome with imputed genotypes.</p>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="phasing.html" class="btn btn-neutral float-left" title="Haplotype Phasing" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="tutorial.html" class="btn btn-neutral float-right" title="Tutorial" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Martin Lab, Broad Institute.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>